### ISSUE TITLE
How to decode images into [0,1]

### ISSUE THREAD

**Issue by tanglang96** [NVIDIA/DALI#1149]
How can I directly decode an image into [0,1], but not [0,255], how can I implement this without normalization ?

**Comment by awolant** [comment]
DALI does not support this. Is there a particular reason why you do not want to use normalization?

**Comment by tanglang96** [comment]
I've compared the same data augmentation(padding, random crop, normalization) implemented by DALI/torchvision on CIFAR10 dataset, and I got a big performance gap(3% accuracy on test set) between them @awolant

**Comment by awolant** [comment]
One of the possible issues is the fact, that DALI uses different libraries to decode images than torchvison does and they are known to produce slightly different results. Make sure you are validating with the data decoded the same way as the data used for training.
Could you provide complete reproduction of both versions, so we could take a closer look and maybe narrow down the issue?
Maybe you will find this [PyTorch issue](https://discuss.pytorch.org/t/normalization-in-the-mnist-example/457/26) helpful.

**Comment by tanglang96** [comment]
this is my DALI implementation
```
class HybridTrainPipe_CIFAR(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, data_dir, crop, dali_cpu=False, local_rank=0, world_size=1,
                 cutout=0):
        super(HybridTrainPipe_CIFAR, self).__init__(batch_size, num_threads, device_id, seed=12 + device_id)
        dali_device = "gpu"
        self.cutout_len = cutout
        self.input = ops.FileReader(file_root=data_dir, shard_id=local_rank, num_shards=world_size, random_shuffle=True)
        self.decode = ops.ImageDecoder(device="mixed", output_type=types.RGB)
        self.pad = ops.Paste(device=dali_device, ratio=1.25, fill_value=0)
        self.uniform = ops.Uniform(range=(0., 1.))
        self.crop = ops.Crop(device=dali_device, crop_h=32, crop_w=32)
        self.cmnp = ops.CropMirrorNormalize(device="gpu",
                                            output_dtype=types.FLOAT,
                                            output_layout=types.NCHW,
                                            image_type=types.RGB,
                                            mean=[125.31, 122.95, 113.87],
                                            std=[63.0, 62.09, 66.70]
                                            )
        self.coin = ops.CoinFlip(probability=0.5)
        self.flip = ops.Flip(device="gpu")

    def define_graph(self):
        rng = self.coin()
        self.jpegs, self.labels = self.input(name="Reader")
        output = self.decode(self.jpegs)
        output = self.pad(output.gpu())
        output = self.crop(output, crop_pos_x=self.uniform(), crop_pos_y=self.uniform())
        output = self.flip(output, horizontal=rng)
        output = self.cmnp(output)
        return [output, self.labels]


class HybridValPipe_CIFAR(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, data_dir, crop, size, local_rank=0, world_size=1):
        super(HybridValPipe_CIFAR, self).__init__(batch_size, num_threads, device_id, seed=12 + device_id)
        self.input = ops.FileReader(file_root=data_dir, shard_id=local_rank, num_shards=world_size,
                                    random_shuffle=False)
        self.decode = ops.ImageDecoder(device="mixed", output_type=types.RGB)
        self.cmnp = ops.CropMirrorNormalize(device="gpu",
                                            output_dtype=types.FLOAT,
                                            output_layout=types.NCHW,
                                            image_type=types.RGB,
                                            mean=[125.31, 122.95, 113.87],
                                            std=[63.0, 62.09, 66.70]
                                            )

    def define_graph(self):
        self.jpegs, self.labels = self.input(name="Reader")
        output = self.decode(self.jpegs)
        output = self.cmnp(output)
        return [output, self.labels]


def get_cifar_iter(type, image_dir, batch_size, num_threads, local_rank=0, world_size=1, val_size=32, cutout=0):
    if type == 'train':
        pip_train = HybridTrainPipe_CIFAR(batch_size=batch_size, num_threads=num_threads, device_id=local_rank,
                                          data_dir=image_dir + '/train',
                                          crop=32, world_size=world_size, local_rank=local_rank, cutout=cutout)
        pip_train.build()
        dali_iter_train = DALIClassificationIterator(pip_train, size=pip_train.epoch_size("Reader") // world_size)
        return dali_iter_train

    elif type == 'val':  
        pip_val = HybridValPipe_CIFAR(batch_size=batch_size, num_threads=num_threads, device_id=local_rank,
                                      data_dir=image_dir + '/test',
                                      crop=32, size=val_size, world_size=world_size, local_rank=local_rank)
        pip_val.build()
        dali_iter_val = DALIClassificationIterator(pip_val, size=pip_val.epoch_size("Reader") // world_size)
        return dali_iter_val`
```
this is my torchvision implementation
```
def get_cifar_iter(type, image_dir, batch_size, num_threads, cutout=0):
    CIFAR_MEAN = [0.49139968, 0.48215827, 0.44653124]
    CIFAR_STD = [0.24703233, 0.24348505, 0.26158768]
    if type == 'train':
        transform_train = transforms.Compose([
            transforms.RandomCrop(32, padding=4),
            transforms.RandomHorizontalFlip(),
            transforms.ToTensor(),
            transforms.Normalize(CIFAR_MEAN, CIFAR_STD),
        ])
        train_dst = CIFAR10(root=image_dir, train=True, download=True, transform=transform_train)
        train_iter = torch.utils.data.DataLoader(train_dst, batch_size=batch_size, shuffle=True, pin_memory=True,
                                                 num_workers=num_threads)
        return train_iter
    else:
        transform_test = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize(CIFAR_MEAN, CIFAR_STD),
        ])
        test_dst = CIFAR10(root=image_dir, train=False, download=True, transform=transform_test)
        test_iter = torch.utils.data.DataLoader(test_dst, batch_size=batch_size, shuffle=False, pin_memory=True,
                                                num_workers=num_threads)
        return test_iter
```
@awolant I hope you can help me check this

**Comment by tanglang96** [comment]
I've fixed this problem, decoding CIFAR10 from `.jpeg` format will lost some information and the problem has been solved after I loaded them from the original binary files

**Comment by samra-irshad** [comment]
> I've fixed this problem, decoding CIFAR10 from `.jpeg` format will lost some information and the problem has been solved after I loaded them from the original binary files

How did you normalize the images to [0,1] range?

**Comment by JanuszL** [comment]
You can use `CropMirrorNormalize` and set std to 255, or just in the define graph step do
```
    output = self.decode(self.jpegs)
    output = output / 255
```
